### PR TITLE
feat: [POC] javascript-based filesystem interface 

### PR DESCRIPTION
**Pull request by rostislavdeepcrawl**

### CODE REVIEW

File: addon/async_executor.cc [hunk 1]
```diff
@@ -8,44 +8,48 @@
...
+      1,                            // Only one thread will use this initially
+      [&](Napi::Env) {              // Finalizer used to clean threads up
+        nativeThread.join();
+      });
+
+  nativeThread = std::thread([&] {
+    auto threadsafe_fn_callback = [&](Napi::Env env,
+                                      Napi::Function js_callback) {
+      if (!result->success) {
+        // TODO: why can't pass Error?
```
Comment (jupiter) [inline addon/async_executor.cc:33]: Do you mean you'd want to pass an Error object?

File: addon/connection.cc [hunk 2]
```diff
@@ -87,30 +93,18 @@ Napi::Value Connection::Execute(const Napi::CallbackInfo &info) {
...
       }
     }
-
-    AsyncExecutor *wk = new AsyncExecutor(env, query, connection, deferred,
-                                          forceMaterializedValue,
-                                          rowResultFormatValue, results);
-    wk->Queue();
+    async_executors->push_back(duckdb::make_unique<NodeDuckDB::AsyncExecutor>(
+        info, query, connection, forceMaterializedValue, rowResultFormatValue));
+    async_executors->back()->Execute();
```
Comment (jupiter) [inline addon/connection.cc:98]: Do we ever clean these up?

File: addon/duckdb.cc [hunk 3]
```diff
@@ -61,62 +60,157 @@ DuckDB::DuckDB(const Napi::CallbackInfo &info)
...
+        nativeThread.join();
+      });
+
+  nativeThread = std::thread([&] {
+    database = duckdb::make_unique<duckdb::DuckDB>(path, &nativeConfig);
+    database->LoadExtension<duckdb::ParquetExtension>();
+    init_tsfn.BlockingCall();
+    init_tsfn.Release();
+  });
+  return info.Env().Undefined();
```
Comment (jupiter) [inline addon/duckdb.cc:141]: Could be nice to return the DuckDB instance here for single line instantiation, e.g. 

```ts
const duckdb = await new DuckDB().init();
```

File: src/tests/perf.test.ts [hunk 4]
```diff
@@ -8,178 +11,85 @@ import { Connection, DuckDB } from "@addon";
...
   let db: DuckDB;
-  let connection: Connection;
+  const path = "";
   beforeEach(async () => {
-    db = new DuckDB();
-    connection = new Connection(db);
-    await connection.executeIterator("PRAGMA threads=4;");
+    db = new DuckDB({ options: { accessMode: AccessMode.ReadWrite, useDirectIO: false } }, fileSystem);
+    await db.init();
+    // await connection.executeIterator("PRAGMA threads=1;");
```
Comment (jupiter) [inline src/tests/perf.test.ts:18]: Does this need to be commented out?

File: package.json [hunk 5]
```diff
@@ -54,11 +54,13 @@
...
     "prettier:fix": "prettier --write '**/*.{js,json,md,ts,yml}'",
     "postpublish": "yarn cleanup:binaries && yarn install && yarn prebuild:upload",
-    "test": "yarn build:ts && NODE_OPTIONS='--max-old-space-size=8192' jest --runInBand --testTimeout=60000"
+    "test": "yarn build:ts && NODE_OPTIONS='--max-old-space-size=8192' node --expose_gc ./node_modules/.bin/jest --runInBand --testTimeout=6000000",
+    "test-valgrind": "yarn build:ts && valgrind --trace-children=yes --log-file='valgrind-2.log' --leak-check=full --show-leak-kinds=all --verbose node --expose_gc ./node_modules/.bin/jest --runInBand --testTimeout=6000000"
   },
   "dependencies": {
     "@microsoft/api-extractor": "^7.12.0",
     "cmake-js": "^6.1.0",
+    "fast-glob": "^3.2.5",
```
Comment (jupiter) [inline package.json:63]: Looks like this is only used in tests, so should be devDependency?

File: addon/async_executor.cc (hunk 1 above)
Comment (rostislavdeepcrawl) [reply addon/async_executor.cc:33]: Yes, but it doesn't compile

File: addon/connection.cc (hunk 2 above)
Comment (rostislavdeepcrawl) [reply addon/connection.cc:98]: connection's destructor should destroy these

File: src/tests/node-filesystem.ts [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import { read, stat, open, ftruncate } from "fs";
 
 import { IFileSystem } from "@addon-types";
+// eslint-disable-next-line node/no-unpublished-import
```
Comment (jupiter) [inline src/tests/node-filesystem.ts:4]: Ah, I see. We probably have this excluded in eslint config for tests in other projects.

**Review by jupiter — APPROVED** [review]
Looks good 👍

**Pull request merged**